libyul/AsmAnalysis.cpp

```diff
@@ -338,12 +338,21 @@ vector<YulString> AsmAnalyzer::operator()(FunctionCall const& _funCall)
 				_funCall.functionName.name.str() == "datasize" ||
 				_funCall.functionName.name.str() == "dataoffset"
 			)
-				if (!m_dataNames.count(std::get<Literal>(arg).value))
+			{
+				auto const lit = std::get<Literal>(arg);
+				if (lit.kind != LiteralKind::String)
```

Comment:
I think we should do that for all literal args, not just these two functions.

Reply:
Not all of them will be strings, though (at least starting with the ``freememorystart`` builtin for https://github.com/ethereum/solidity/pull/9162/).
Maybe when defining literal arguments it shouldn't be done by a vector of ``bool``, but by a vector of ``optional<LiteralKind>``s and those should be verified here?

Reply:
True! Forgot about that one! Your idea sounds good!

Reply:
What is the suggestion then?

Reply:
https://github.com/ethereum/solidity/pull/9564#discussion_r464878951 This :-)?

Reply:
I.e. replace the ``vector<bool>`` ``needsLiteralArguments`` thing in the builtins with a ``vector<optional<LiteralKind>>`` (and when defining the builtins change e.g. ``{true}`` to ``{{LiteralKind::String}}`` for ``datasize``, etc.).
Then instead of verifying that each argument for which this vector has a ``true`` entry is a string literal, verify that each argument with a non-empty ``optional`` in that vector has a literal of the ``LiteralKind`` given by that optional as literal argument.